README.md
# go-audit

[![License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](http://opensource.org/licenses/MIT)

## About

go-audit is an alternative to the auditd daemon that ships with many distros.
After having created an [auditd audisp](https://people.redhat.com/sgrubb/audit/) plugin to convert audit logs to json, 
I became interested in creating a replacement for the existing daemon.

##### Goals

* Safe : Written in a modern language that is type safe and performant
* Fast : Never ever ever ever block if we can avoid it
* Outputs json : Yay
* Pluggable pipelines : Can write to syslog, local file, Graylog2 or stdout. Additional outputs are easily written. 
* Connects to the linux kernel via netlink (info [here](https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/tree/kernel/audit.c?id=refs/tags/v3.14.56) and [here](https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/tree/include/uapi/linux/audit.h?h=linux-3.14.y))

## Usage

##### Installation

1. Install [golang](https://golang.org/doc/install), version 1.14 or greater is required
2. Clone the repo

    ```
    git clone (this repo)
    cd go-audit
    ```
    
3. Build the binary

    ```
    make
    ```

4. Copy the binary `go-audit` to wherever you'd like

##### Testing

- `make test` - run the unit test suite
- `make test-cov-html` - run the unit tests and open up the code coverage results
- `make bench` - run the benchmark test suite
- `make bench-cpu` - run the benchmark test suite with cpu profiling
- `make bench-cpulong` - run the benchmark test suite with cpu profiling and try to get some gc collection

##### Running as a service
 
Check the [contrib](contrib) folder, it contains examples for how to run `go-audit` as a proper service on your machine.

##### Example Config 

See [go-audit.yaml.example](go-audit.yaml.example)

## FAQ

#### I am seeing `Error during message receive: no buffer space available` in the logs

This is because `go-audit` is not receiving data as quickly as your system is generating it. You can increase
the receive buffer system wide and maybe it will help. Best to try and reduce the amount of data `go-audit` has
to handle.

If reducing audit velocity is not an option you can try increasing `socket_buffer.receive` in your config.
See [Example Config](#example-config) for more information

```
socket_buffer:
    receive: <some number bigger than (the current value * 2)>
```

#### Sometime files don't have a `name`, only `inode`, what gives?

The kernel doesn't always know the filename for file access. Figuring out the filename from an inode is expensive and
error prone.

You can map back to a filename, possibly not *the* filename, that triggured the audit line though.

```
sudo debugfs -R "ncheck <inode to map>" /dev/<your block device here>
```

#### I don't like math and want you to tell me the syslog priority to use

Use the default, or consult this handy table.

Wikipedia has a pretty good [page](https://en.wikipedia.org/wiki/Syslog) on this

|                   | emerg (0)| alert (1) | crit (2)  | err (3) | warn (4) | notice (5) | info (6)  | debug (7) |
|-------------------|----------|-----------|-----------|---------|----------|------------|-----------|-----------|
| **kernel (0)**    | 0        | 1         | 2         | 3       | 4        | 5          | 6         | 7         |
| **user (1)**      | 8        | 9         | 10        | 11      | 12       | 13         | 14        | 15        |
| **mail (2)**      | 16       | 17        | 18        | 19      | 20       | 21         | 22        | 23        |
| **daemon (3)**    | 24       | 25        | 26        | 27      | 28       | 29         | 30        | 31        |
| **auth (4)**      | 32       | 33        | 34        | 35      | 36       | 37         | 38        | 39        |
| **syslog (5)**    | 40       | 41        | 42        | 43      | 44       | 45         | 46        | 47        |
| **lpr (6)**       | 48       | 49        | 50        | 51      | 52       | 53         | 54        | 55        |
| **news (7)**      | 56       | 57        | 58        | 59      | 60       | 61         | 62        | 63        |
| **uucp (8)**      | 64       | 65        | 66        | 67      | 68       | 69         | 70        | 71        |
| **clock (9)**     | 72       | 73        | 74        | 75      | 76       | 77         | 78        | 79        |
| **authpriv (10)** | 80       | 81        | 82        | 83      | 84       | 85         | 86        | 87        |
| **ftp (11)**      | 88       | 89        | 90        | 91      | 92       | 93         | 94        | 95        |
| **ntp (12)**      | 96       | 97        | 98        | 99      | 100      | 101        | 102       | 103       |
| **logaudit (13)** | 104      | 105       | 106       | 107     | 108      | 109        | 110       | 111       |
| **logalert (14)** | 112      | 113       | 114       | 115     | 116      | 117        | 118       | 119       |
| **cron (15)**     | 120      | 121       | 122       | 123     | 124      | 125        | 126       | 127       |
| **local0 (16)**   | 128      | 129       | 130       | 131     | 132      | 133        | 134       | 135       |
| **local1 (17)**   | 136      | 137       | 138       | 139     | 140      | 141        | 142       | 143       |
| **local2 (18)**   | 144      | 145       | 146       | 147     | 148      | 149        | 150       | 151       |
| **local3 (19)**   | 152      | 153       | 154       | 155     | 156      | 157        | 158       | 159       |
| **local4 (20)**   | 160      | 161       | 162       | 163     | 164      | 165        | 166       | 167       |
| **local5 (21)**   | 168      | 169       | 170       | 171     | 172      | 173        | 174       | 175       |
| **local6 (22)**   | 176      | 177       | 178       | 179     | 180      | 181        | 182       | 183       |
| **local7 (23)**   | 184      | 185       | 186       | 187     | 188      | 189        | 190       | 191       |

#### I am seeing duplicate entries in syslog!

This is likely because you are running `journald` which is also reading audit events. To disable it you need to disable the functionality in `journald`.

```sh
sudo systemctl mask systemd-journald-audit.socket
```

## Thanks!

To Hardik Juneja, Arun Sori, Aalekh Nigam Aalekhn for the inspiration via https://github.com/mozilla/audit-go


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2023-04-07

### Added

- Add `containerd` support in the same way we support grabbing container
  metadata from docker. (#106)

- Build releases for arm64. (#107)

## [1.1.1] - 2022-01-24

### Fixed

- Fix `cgroup` support on non-amd64 architectures. (#96)

## [1.1.0] - 2022-01-24

### Added

- Added `cgroup` optional parser. When enabled, this will annotate audit
  events with the cgroup v2 root path. If running on an older system with only
  cgroup v1, this falls back to the pid cgroup path. (#95)

## [1.0.0] - 2020-06-18

### Added

- You can now run `go-audit -version` to check the build version.

[Unreleased]: https://github.com/slackhq/go-audit/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/slackhq/go-audit/releases/tag/v1.2.0
[1.1.1]: https://github.com/slackhq/go-audit/releases/tag/v1.1.1
[1.1.0]: https://github.com/slackhq/go-audit/releases/tag/v1.1.0
[1.0.0]: https://github.com/slackhq/go-audit/releases/tag/v1.0.0


audit.go
package main

import (
	"compress/flate"
	"errors"
	"flag"
	"fmt"
	"log"
	"log/syslog"
	"os"
	"os/exec"
	"os/signal"
	"os/user"
	"regexp"
	"strconv"
	"strings"
	"syscall"

	"github.com/spf13/viper"
	"gopkg.in/Graylog2/go-gelf.v2/gelf"
)

// A version string that can be set with
//
//	-ldflags "-X main.Build=SOMEVERSION"
//
// at compile-time.
var Build string

var l = log.New(os.Stdout, "", 0)
var el = log.New(os.Stderr, "", 0)

type executor func(string, ...string) error

func lExec(s string, a ...string) error {
	return exec.Command(s, a...).Run()
}

func loadConfig(configFile string) (*viper.Viper, error) {
	config := viper.New()
	config.SetConfigFile(configFile)

	config.SetDefault("events.min", 1300)
	config.SetDefault("events.max", 1399)
	config.SetDefault("message_tracking.enabled", true)
	config.SetDefault("message_tracking.log_out_of_order", false)
	config.SetDefault("message_tracking.max_out_of_order", 500)
	config.SetDefault("output.syslog.enabled", false)
	config.SetDefault("output.syslog.priority", int(syslog.LOG_LOCAL0|syslog.LOG_WARNING))
	config.SetDefault("output.syslog.tag", "go-audit")
	config.SetDefault("output.syslog.attempts", "3")
	config.SetDefault("output.gelf.attempts", 3)
	config.SetDefault("output.gelf.network", "udp")
	config.SetDefault("output.gelf.compression.level", int(flate.BestSpeed))
	config.SetDefault("output.gelf.compression.type", int(gelf.CompressGzip))
	config.SetDefault("log.flags", 0)

	if err := config.ReadInConfig(); err != nil {
		return nil, err
	}

	l.SetFlags(config.GetInt("log.flags"))
	el.SetFlags(config.GetInt("log.flags"))

	return config, nil
}

func setRules(config *viper.Viper, e executor) error {
	// Clear existing rules
	if err := e("auditctl", "-D"); err != nil {
		return fmt.Errorf("Failed to flush existing audit rules. Error: %s", err)
	}

	l.Println("Flushed existing audit rules")

	// Add ours in
	if rules := config.GetStringSlice("rules"); len(rules) != 0 {
		for i, v := range rules {
			// Skip rules with no content
			if v == "" {
				continue
			}

			if err := e("auditctl", strings.Fields(v)...); err != nil {
				return fmt.Errorf("Failed to add rule #%d. Error: %s", i+1, err)
			}

			l.Printf("Added audit rule #%d\n", i+1)
		}
	} else {
		return errors.New("No audit rules found")
	}

	return nil
}

func createOutput(config *viper.Viper) (*AuditWriter, error) {
	var writer *AuditWriter
	var err error
	i := 0

	if config.GetBool("output.syslog.enabled") == true {
		i++
		writer, err = createSyslogOutput(config)
		if err != nil {
			return nil, err
		}
	}

	if config.GetBool("output.file.enabled") == true {
		i++
		writer, err = createFileOutput(config)
		if err != nil {
			return nil, err
		}

		go handleLogRotation(config, writer)
	}

	if config.GetBool("output.stdout.enabled") == true {
		i++
		writer, err = createStdOutOutput(config)
		if err != nil {
			return nil, err
		}
	}

	if config.GetBool("output.gelf.enabled") == true {
		i++
		writer, err = createGELFOutput(config)
		if err != nil {
			return nil, err
		}
	}

	if i > 1 {
		return nil, errors.New("Only one output can be enabled at a time")
	}

	if writer == nil {
		return nil, errors.New("No outputs were configured")
	}

	return writer, nil
}

func createGELFOutput(config *viper.Viper) (*AuditWriter, error) {
	attempts := config.GetInt("output.gelf.attempts")
	if attempts < 1 {
		return nil, fmt.Errorf("Output attempts for GELF must be at least 1, %v provided", attempts)
	}

	address := config.GetString("output.gelf.address")
	if address == "" {
		return nil, fmt.Errorf("Output address for GELF must be set")
	}

	switch config.GetString("output.gelf.network") {
	case "udp":
		writer, err := gelf.NewUDPWriter(address)
		if err != nil {
			return nil, err
		}

		writer.CompressionType = gelf.CompressType(config.GetInt("output.gelf.compression.type"))
		writer.CompressionLevel = config.GetInt("output.gelf.compression.level")

		return NewAuditWriter(writer, attempts), nil
	case "tcp":
		writer, err := gelf.NewTCPWriter(address)
		if err != nil {
			return nil, err
		}

		return NewAuditWriter(writer, attempts), nil
	default:
		return nil, fmt.Errorf("unsupported network by GELF library")
	}

}

func createSyslogOutput(config *viper.Viper) (*AuditWriter, error) {
	attempts := config.GetInt("output.syslog.attempts")
	if attempts < 1 {
		return nil, fmt.Errorf("Output attempts for syslog must be at least 1, %v provided", attempts)
	}

	syslogWriter, err := syslog.Dial(
		config.GetString("output.syslog.network"),
		config.GetString("output.syslog.address"),
		syslog.Priority(config.GetInt("output.syslog.priority")),
		config.GetString("output.syslog.tag"),
	)

	if err != nil {
		return nil, fmt.Errorf("Failed to open syslog writer. Error: %v", err)
	}

	return NewAuditWriter(syslogWriter, attempts), nil
}

func createFileOutput(config *viper.Viper) (*AuditWriter, error) {
	attempts := config.GetInt("output.file.attempts")
	if attempts < 1 {
		return nil, fmt.Errorf("Output attempts for file must be at least 1, %v provided", attempts)
	}

	mode := os.FileMode(config.GetInt("output.file.mode"))
	if mode < 1 {
		return nil, errors.New("Output file mode should be greater than 0000")
	}

	f, err := os.OpenFile(
		config.GetString("output.file.path"),
		os.O_APPEND|os.O_CREATE|os.O_WRONLY, mode,
	)

	if err != nil {
		return nil, fmt.Errorf("Failed to open output file. Error: %s", err)
	}

	if err := f.Chmod(mode); err != nil {
		return nil, fmt.Errorf("Failed to set file permissions. Error: %s", err)
	}

	uname := config.GetString("output.file.user")
	u, err := user.Lookup(uname)
	if err != nil {
		return nil, fmt.Errorf("Could not find uid for user %s. Error: %s", uname, err)
	}

	gname := config.GetString("output.file.group")
	g, err := user.LookupGroup(gname)
	if err != nil {
		return nil, fmt.Errorf("Could not find gid for group %s. Error: %s", gname, err)
	}

	uid, err := strconv.ParseInt(u.Uid, 10, 32)
	if err != nil {
		return nil, fmt.Errorf("Found uid could not be parsed. Error: %s", err)
	}

	gid, err := strconv.ParseInt(g.Gid, 10, 32)
	if err != nil {
		return nil, fmt.Errorf("Found gid could not be parsed. Error: %s", err)
	}

	if err = f.Chown(int(uid), int(gid)); err != nil {
		return nil, fmt.Errorf("Could not chown output file. Error: %s", err)
	}

	return NewAuditWriter(f, attempts), nil
}

func handleLogRotation(config *viper.Viper, writer *AuditWriter) {
	// Re-open our log file. This is triggered by a USR1 signal and is meant to be used upon log rotation

	sigc := make(chan os.Signal, 1)
	signal.Notify(sigc, syscall.SIGUSR1)

	for range sigc {
		newWriter, err := createFileOutput(config)
		if err != nil {
			el.Fatalln("Error re-opening log file. Exiting.")
		}

		oldFile := writer.w.(*os.File)
		writer.w = newWriter.w
		writer.e = newWriter.e

		err = oldFile.Close()
		if err != nil {
			el.Printf("Error closing old log file: %+v\n", err)
		}
	}
}

func createStdOutOutput(config *viper.Viper) (*AuditWriter, error) {
	attempts := config.GetInt("output.stdout.attempts")
	if attempts < 1 {
		return nil, fmt.Errorf("Output attempts for stdout must be at least 1, %v provided", attempts)
	}

	// l logger is no longer stdout
	l.SetOutput(os.Stderr)

	return NewAuditWriter(os.Stdout, attempts), nil
}

func createFilters(config *viper.Viper) ([]AuditFilter, error) {
	var err error
	var ok bool

	fs := config.Get("filters")
	filters := []AuditFilter{}

	if fs == nil {
		return filters, nil
	}

	ft, ok := fs.([]interface{})
	if !ok {
		return filters, fmt.Errorf("Could not parse filters object")
	}

	for i, f := range ft {
		f2, ok := f.(map[string]interface{})
		if !ok {
			return filters, fmt.Errorf("Could not parse filter %d; '%+v'", i+1, f)
		}

		af := AuditFilter{}
		for k, v := range f2 {
			switch k {
			case "message_type":
				if ev, ok := v.(string); ok {
					fv, err := strconv.ParseUint(ev, 10, 64)
					if err != nil {
						return filters, fmt.Errorf("`message_type` in filter %d could not be parsed; Value: `%+v`; Error: %s", i+1, v, err)
					}
					af.messageType = uint16(fv)

				} else if ev, ok := v.(int); ok {
					af.messageType = uint16(ev)

				} else {
					return filters, fmt.Errorf("`message_type` in filter %d could not be parsed; Value: `%+v`", i+1, v)
				}

			case "regex":
				re, ok := v.(string)
				if !ok {
					return filters, fmt.Errorf("`regex` in filter %d could not be parsed; Value: `%+v`", i+1, v)
				}

				if af.regex, err = regexp.Compile(re); err != nil {
					return filters, fmt.Errorf("`regex` in filter %d could not be parsed; Value: `%+v`; Error: %s", i+1, v, err)
				}

			case "syscall":
				if af.syscall, ok = v.(string); ok {
					// All is good
				} else if ev, ok := v.(int); ok {
					af.syscall = strconv.Itoa(ev)
				} else {
					return filters, fmt.Errorf("`syscall` in filter %d could not be parsed; Value: `%+v`", i+1, v)
				}
			}
		}

		if af.regex == nil {
			return filters, fmt.Errorf("Filter %d is missing the `regex` entry", i+1)
		}

		if af.messageType == 0 {
			return filters, fmt.Errorf("Filter %d is missing the `message_type` entry", i+1)
		}

		filters = append(filters, af)
		l.Printf("Ignoring syscall `%v` containing message type `%v` matching string `%s`\n", af.syscall, af.messageType, af.regex.String())
	}

	return filters, nil
}

func main() {
	configFile := flag.String("config", "", "Config file location")
	printVersion := flag.Bool("version", false, "Print version")

	flag.Parse()

	if *printVersion {
		fmt.Printf("Version: %s\n", Build)
		os.Exit(0)
	}

	if *configFile == "" {
		el.Println("A config file must be provided")
		flag.Usage()
		os.Exit(1)
	}

	config, err := loadConfig(*configFile)
	if err != nil {
		el.Fatal(err)
	}

	// output needs to be created before anything that write to stdout
	writer, err := createOutput(config)
	if err != nil {
		el.Fatal(err)
	}

	if err := setRules(config, lExec); err != nil {
		el.Fatal(err)
	}

	filters, err := createFilters(config)
	if err != nil {
		el.Fatal(err)
	}

	nlClient, err := NewNetlinkClient(config.GetInt("socket_buffer.receive"))
	if err != nil {
		el.Fatal(err)
	}

	marshaller := NewAuditMarshaller(
		writer,
		uint16(config.GetInt("events.min")),
		uint16(config.GetInt("events.max")),
		config.GetBool("message_tracking.enabled"),
		config.GetBool("message_tracking.log_out_of_order"),
		config.GetInt("message_tracking.max_out_of_order"),
		filters,
		createExtraParsers(config),
	)

	l.Printf("Started processing events in the range [%d, %d]\n", config.GetInt("events.min"), config.GetInt("events.max"))

	//Main loop. Get data from netlink and send it to the json lib for processing
	for {
		msg, err := nlClient.Receive()
		if err != nil {
			el.Printf("Error during message receive: %+v\n", err)
			continue
		}

		if msg == nil {
			continue
		}

		marshaller.Consume(msg)
	}
}
